### PR TITLE
add Docker Swarm support

### PR DESCRIPTION
**Pull request by s4ke**
ready for review

closes https://github.com/hetznercloud/csi-driver/issues/374

### CODE REVIEW

**Comment by s4ke** [comment]
@apricote I think this is ready for review. The only thing missing right now is the build pipeline. For this I was wondering whether we should simply reuse the Makefile but wrap it in a GitHub action. What do you think?

File: driver/node.go [hunk 1]
```diff
@@ -164,6 +166,13 @@ func (s *NodeService) NodeGetCapabilities(ctx context.Context, req *proto.NodeGe
 					},
 				},
 			},
+			{
+				Type: &proto.NodeServiceCapability_Rpc{
```
Comment (apricote) [inline driver/node.go:170]: Adding this capability adds additional gRPC calls made for volumes in other Container Orchestrators. I would prefer to wait for moby/swarmkit#3116 as Swarm is not csi-compliant right now.

File: deploy/docker-swarm/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
+# Docker Swarm Hetzner CSI plugin
+
+Currently in Beta. Please consult the Docker Swarm documentation
```
Comment (apricote) [inline deploy/docker-swarm/README.md:3]: What exactly is in `beta`? The _Docker Swarm CSI_ support or the _hcloud-csi-driver Docker Swarm_ support?

File: deploy/docker-swarm/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+Run the following steps on all nodes (especially master nodes).
+The simplest way to achieve this
+
+1. Install the plugin 
+
+```bash
+docker plugin install --disable --alias hetznercloud/csi-driver-docker --grant-all-permissions hetznercloud/csi-driver-docker
+```
+
+2. Set HCLOUD_TOKEN
```
Comment (apricote) [inline deploy/docker-swarm/README.md:17]: Could you describe how the user needs to create the token and what permissions are required?

See in our kubernetes guide: https://github.com/hetznercloud/csi-driver/blob/e445d79e0a22fc5637e8862e76e0160bb20983ff/README.md?plain=1#L12

File: deploy/docker-swarm/README.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+After creation we can now use this volume with `--sharing all` in more than one replica:
+
+```bash
+docker service create --name hcloud-debug-serv2  --mount type=cluster,src=hcloud-debug2,dst=/srv/www   nginx:alpine
+docker service scale hcloud-debug-serv2=2
+```
+
+## How to resize a docker swarm Hetzner CSI volume
+
+Currently, the Docker Swarm CSI support does not come with support for volume resizing.
```
Comment (apricote) [inline deploy/docker-swarm/README.md:60]: Is there an upstream ticket that we can link here so users can check if support has been added by the time they read this guide?

File: deploy/docker-swarm/README.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+4. Resize Volume in Hetzner UI
+5. Attach Volume to temporary server manually
+6. Run resize2fs manually
+7. Detach Volume from temporary server manually
+8. Recreate Volume with new size to make it known to Swarm again
+
+```
+docker volume create   --driver hetznercloud/csi-driver-docker:dev --required-bytes <new-size>  --type mount   --sharing onewriter   --scope single <volume-name>
```
Comment (apricote) [inline deploy/docker-swarm/README.md:89]: This uses `--driver hetznercloud/csi-driver-docker:dev`, but all other examples did not specify `:dev`

File: deploy/docker-swarm/pkg/Makefile [hunk 6]
```diff
@@ -0,0 +1,32 @@
+# FIXME: change this
+PLUGIN_NAME = hetznercloud/csi-driver-docker
```
Comment (apricote) [inline deploy/docker-swarm/pkg/Makefile:2]: I would prefer to use the same docker image as for the kubernetes(/general) csi support: `hetznercloud/hcloud-csi-driver`. To differentiate between the normal and Docker Swarm images, we can use tags:

- `hetznercloud/hcloud-csi-driver:latest` -> normal
- `hetznercloud/hcloud-csi-driver:latest-swarm` -> Docker Swarm

File: cmd/aio/main.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+		os.Exit(1)
+	}
+	parts := strings.Split(serverAZ, "-")
+	if len(parts) != 2 {
+		level.Error(logger).Log("msg", fmt.Sprintf("unexpected server availability zone: %s", serverAZ), "err", err)
+		os.Exit(1)
+	}
+	serverLocation := parts[0]
+
+	level.Info(logger).Log("msg", "Fetched data from metadata service", "id", serverID, "location", serverLocation)
```
Comment (apricote) [inline cmd/aio/main.go:65]: Lines 37-44 are from `cmd/controller` while 46-65 are from `cmd/node`. Both of these retrieve the server metadata.

You can remove one of these and just use the return value of the other to replace any usages.

E.g. you could replace

- `server.Datacenter.Location.Name` -> `serverLocation`

Or the other way round:

- `serverID` -> `server.ID`
- `serverLocation` -> `server.Datacenter.Location.Name`

## Background

We have these two differing implementations as the both serve different niches:

- The controller implementation should not rely on the metadata service, as the controller part may run on non-hcloud servers
- The node implementation may not use the Hetzner Cloud api, as access to it is not necessary and we do not need to distribute the API access token to all nodes this way

As this binary needs to access the hcloud API AND the metadata service anyway, it does not really matter which one you choose.

**Comment by apricote** [comment]
>I think this is ready for review.

Done :) Thanks for all the work you are putting into this!

> The only thing missing right now is the build pipeline. For this I was wondering whether we should simply reuse the Makefile but wrap it in a GitHub action. What do you think?

That sounds good to me, you need to add it the the workflows `publish_on_master.yml` and `publish_on_tag.yml`

File: deploy/docker-swarm/pkg/config.json [hunk 8]
```diff
@@ -0,0 +1,62 @@
+{
+  "description": "Hetzner csi-driver plugin for Docker",
+  "documentation": "https://github.com/hetznercloud/csi-driver",
+  "entrypoint": [
+    "/plugin/aio.bin"
```
[the remark this answers is not in the document]
Comment (s4ke) [reply deploy/docker-swarm/pkg/config.json:5]: We have to package the docker plugin in a special way that is actually different to regular Docker image builds. Essentially the Dockerfile in the pkg directory is just the "builder" that we then harness for its rootfs which we use to create the plugin out of.

File: deploy/docker-swarm/README.md (hunk 2 above)
Comment (s4ke) [reply deploy/docker-swarm/README.md:3]: Good that you point this out.

According to https://docs.docker.com/engine/release-notes/23.0/ the CSI support is "experimental" but I don't  think this is to be read as "may be removed again".

I added this line to this file to mention that there is no guarantee that this plugin + Docker Swarm CSI will work 100% for everyone. This is a new feature after all.

File: deploy/docker-swarm/README.md (hunk 4 above)
Comment (s4ke) [reply deploy/docker-swarm/README.md:60]: I just created one https://github.com/moby/moby/issues/44985

File: deploy/docker-swarm/pkg/Makefile (hunk 6 above)
Comment (s4ke) [reply deploy/docker-swarm/pkg/Makefile:2]: Due to the nature of how docker plugins are packaged, I don't think that this is possible or even something we would want. See https://github.com/hetznercloud/csi-driver/pull/376#discussion_r1104137481

File: driver/node.go (hunk 1 above)
Comment (s4ke) [reply driver/node.go:170]: Hmm, as an alternative, would you be fine with hiding this behind a feature flag that we can enable/disable via an ENV var that we automatically set in the Docker plugin?

I totally understand your point though and once the PR is merged, I would be happy to provide another PR that removes the flag.

File: deploy/docker-swarm/README.md (hunk 3 above)
Comment (s4ke) [reply deploy/docker-swarm/README.md]: I added a similar line

File: deploy/docker-swarm/README.md (hunk 4 above)
Comment (s4ke) [reply deploy/docker-swarm/README.md]: done

File: deploy/docker-swarm/pkg/Makefile [hunk 9]
```diff
@@ -0,0 +1,32 @@
+# FIXME: change this
```
Comment (apricote) [inline deploy/docker-swarm/pkg/Makefile:1]: Do you want to address this before merging?

File: deploy/docker-swarm/pkg/config.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (s4ke) [reply deploy/docker-swarm/pkg/config.json:5]: I guess this conversation is the last one left. Is this fine to keep this way or should we change something here? We could maybe harmonize the path to /csi/aio.bin ?

[the remark this answers is not in the document]
Comment (apricote) [reply deploy/docker-swarm/pkg/config.json:5]: This was mostly about re-using the same dockerfile. As we decided against that, I will resolve the conversation.

File: deploy/docker-swarm/pkg/Makefile [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	docker plugin create ${PLUGIN_NAME}:${PLUGIN_TAG} ./plugin
+
+enable: create
+	docker plugin enable ${PLUGIN_NAME}:${PLUGIN_TAG}
+
+push: create
+	docker plugin push ${PLUGIN_NAME}:${PLUGIN_TAG}
+
+push_latest: create
+	docker plugin push ${PLUGIN_NAME}:latest
```
Comment (apricote) [inline deploy/docker-swarm/pkg/Makefile:29]:
```suggestion
	docker plugin push ${PLUGIN_NAME}:latest-swarm
```

**Review by apricote — APPROVED** [review]

**Comment by apricote** [comment]
@s4ke I will test this tomorrow and if everything looks good I will merge it :)

File: driver/node.go [hunk 11]
```diff
@@ -36,17 +42,20 @@ func NewNodeService(
...
-	return nil, status.Error(codes.Unimplemented, "not supported")
+	// while we dont do anything here, Swarm 23.0.1 might require this
+	return &proto.NodeStageVolumeResponse{}, nil
 }
 
 func (s *NodeService) NodeUnstageVolume(ctx context.Context, req *proto.NodeUnstageVolumeRequest) (*proto.NodeUnstageVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "not supported")
+	// while we dont do anything here, Swarm 23.0.1 might require this
+	return &proto.NodeUnstageVolumeResponse{}, nil
 }
```
Comment (neersighted) [inline driver/node.go:59]: I would avoid implementing a hack for a bug in Moby/SwarmKit; https://github.com/moby/swarmkit/pull/3116 will address this.

Comment (s4ke) [reply driver/node.go:59]: @neersighted assuming the PR will take some time to be merged / a release will take a while, in this case and if @apricote is fine with it, I would like to still merge it tbh. This way we can get some real world experience with this driver even though it is not perfect yet. The capabilities are hidden behind a featureflag that does not impact anyone besides new Swarm users.

Once the PR is merged, I will happily create another PR to remove the hack. I totally understand if we don't want to go ahead with this though, as this is not my decision to make :)

**Comment by apricote** [comment]
I am unable to successfully mount a volume to a service in my testing.

## Reproduction Steps

1. Checkout PR
2. Build and push image
   ```shell
   cd deploy/docker-swarm/pkg
   make push PLUGIN_NAME=apricote/hetznercloud_hcloud-csi-driver
   ```
3. Setup server and swarm
   ```shell
   ## prepare server
   hcloud server create --name csi-test-swarm --image ubuntu-22.04 --ssh-key julian.toelle --type cpx21
   hcloud server ssh csi-test-swarm
   # on server
   ## init swarm
   curl -fsSL https://get.docker.com -o get-docker.sh
   sh get-docker.sh
   docker swarm init
   ```
4. Setup plugin (according to docs)
    ```shell
   docker plugin install --disable --alias hetznercloud/hcloud-csi-driver --grant-all-permissions apricote/hetznercloud_hcloud-csi-driver:dev-swarm
   docker plugin set hetznercloud/hcloud-csi-driver HCLOUD_TOKEN=<my token>
   docker plugin enable hetznercloud/hcloud-csi-driver
   ```
5. Create Volume
   ```shell
   docker volume create --driver hetznercloud/hcloud-csi-driver --required-bytes 50G --type mount --sharing onewriter --scope single hcloud-debug1 --topology-required csi.hetzner.cloud/location=hel1
   ```
6. Create Service
   ```shell
   docker service create --name hcloud-debug-serv1   --mount type=cluster,src=hcloud-debug1,dst=/srv/www   nginx:alpine
   ```

The last step hangs indefinitely. The log output (`journalctl -u docker.service -f`) shows that the call to `ControllerPublishVolume` fails:

```
Feb 14 11:31:03 csi-test-swarm dockerd[2056]: time="2023-02-14T11:31:03Z" level=info msg="level=debug ts=2023-02-14T11:31:03.343496798Z component=grpc-server msg=\"handling request\" method=/csi.v1.Controller/ControllerPublishVolume req=\"volume_id:\\\"28096177\\\" volume_capability:<mount:<> access_mode:<mode:SINGLE_NODE_WRITER > > \"" plugin=7cd4cf53fdbf4b60ffbfc2933311183f6bff5164492e40039330156033b6159e
Feb 14 11:31:03 csi-test-swarm dockerd[2056]: time="2023-02-14T11:31:03Z" level=info msg="level=error ts=2023-02-14T11:31:03.343618005Z component=grpc-server msg=\"handler failed\" err=\"rpc error: code = InvalidArgument desc = missing node id\"" plugin=7cd4cf53fdbf4b60ffbfc2933311183f6bff5164492e40039330156033b6159e
Feb 14 11:31:03 csi-test-swarm dockerd[2056]: time="2023-02-14T11:31:03.344369531Z" level=info msg="error handling volume" attempt=0 error="error publishing or unpublishing to some nodes: [o1hvbglvshwwcstm6zzujcyyb]" module=csi/manager node.id=o1hvbglvshwwcstm6zzujcyyb volume.id=kv91gibx6xhtntntq7cqs9gjp
```

## Further Info

Node ID was correctly retrieved from metadata:

```
Feb 14 11:28:06 csi-test-swarm dockerd[2056]: time="2023-02-14T11:28:06Z" level=info msg="level=info ts=2023-02-14T11:28:06.607759961Z msg=\"Fetched data from metadata service\" id=28867925 location=hel1" plugin=7cd4cf53fdbf4b60ffbfc2933311183f6bff5164492e40039330156033b6159e
```

Node has correct info set:

```shell
root@csi-test-swarm:~# docker node inspect o1hvbglvshwwcstm6zzujcyyb
[
    {
        "ID": "o1hvbglvshwwcstm6zzujcyyb",
        ...
        "CreatedAt": "2023-02-14T11:28:42.604637918Z",
        "UpdatedAt": "2023-02-14T11:28:43.209672806Z",
        ...
        "Description": {
            ...,
            "Engine": {
                "EngineVersion": "23.0.1",
                "Plugins": [
                    ...
                    {
                        "Type": "csicontroller",
                        "Name": "hetznercloud/hcloud-csi-driver:latest"
                    },
                    {
                        "Type": "csinode",
                        "Name": "hetznercloud/hcloud-csi-driver:latest"
                    }
                ]
            },
            ...
            "CSIInfo": [
                {
                    "PluginName": "hetznercloud/hcloud-csi-driver:latest",
                    "NodeID": "28867925",
                    "MaxVolumesPerNode": 16
                }
            ]
        },
        ...
    }
]
```

Volume is in state `pending-publish`

```
root@csi-test-swarm:~# docker volume inspect kv91gibx6xhtntntq7cqs9gjp
[
    {
        "ClusterVolume": {
            "ID": "kv91gibx6xhtntntq7cqs9gjp",
            "Version": {
                "Index": 35
            },
            "CreatedAt": "2023-02-14T11:28:48.7175303Z",
            "UpdatedAt": "2023-02-14T11:54:42.381566553Z",
            "Spec": {
                "AccessMode": {
                    "Scope": "single",
                    "Sharing": "onewriter",
                    "MountVolume": {}
                },
                "AccessibilityRequirements": {
                    "Requisite": [
                        {
                            "Segments": {
                                "csi.hetzner.cloud/location": "hel1"
                            }
                        }
                    ]
                },
                "CapacityRange": {
                    "RequiredBytes": 53687091200,
                    "LimitBytes": 0
                },
                "Availability": "active"
            },
            "PublishStatus": [
                {
                    "NodeID": "o1hvbglvshwwcstm6zzujcyyb",
                    "State": "pending-publish"
                }
            ],
            "Info": {
                "CapacityBytes": 53687091200,
                "VolumeID": "28096177",
                "AccessibleTopology": [
                    {
                        "Segments": {
                            "csi.hetzner.cloud/location": "hel1"
                        }
                    }
                ]
            }
        },
        "CreatedAt": "2023-02-14 11:28:48.7175303 +0000 UTC",
        "Driver": "hetznercloud/hcloud-csi-driver",
        "Labels": null,
        "Mountpoint": "",
        "Name": "hcloud-debug1",
        "Options": null,
        "Scope": "global"
    }
]
```

**Comment by s4ke** [comment]
@apricote thanks for the detailed reproduction Steps. I will check what is wrong.

**Comment by s4ke** [comment]
Here are some notes from debugging on my end. There seems to be some issue regarding CSI scheduling that occurs whenever you get into a broken state by creating a volume in a different location than your service is in. This then affects all volumes somehow and we have to clean the state by force removing all volumes and start from scratch.

In any case, this does not seem to be an error of the CSI plugin.

```bash
# working plugin, force removed everything, didnt restart docker

docker plugin install --disable --alias hetznercloud/hcloud-csi-driver --grant-all-permissions apricote/hetznercloud_hcloud-csi-driver:dev-swarm

docker plugin set hetznercloud/hcloud-csi-driver HCLOUD_TOKEN=<token>

docker plugin enable hetznercloud/hcloud-csi-driver

docker volume create --driver hetznercloud/hcloud-csi-driver --required-bytes 11G --type mount --sharing onewriter --scope single hcloud-nbg1 --topology-required csi.hetzner.cloud/location=nbg1
docker volume create --driver hetznercloud/hcloud-csi-driver --required-bytes 11G --type mount --sharing onewriter --scope single hcloud-hel1 --topology-required csi.hetzner.cloud/location=hel1

# pending creation (had plugin installed before, have to check with Docker team if this is expected behaviour)

martinb@csi-dev02:~/csi-driver$ docker volume ls --cluster
VOLUME NAME   GROUP     DRIVER                           AVAILABILITY   STATUS
hcloud-hel1             hetznercloud/hcloud-csi-driver   active         pending creation
hcloud-nbg1             hetznercloud/hcloud-csi-driver   active         pending creation

# restarting docker
# as root:
service docker restart

# creation worked
martinb@csi-dev02:~/csi-driver$ docker volume ls --cluster
VOLUME NAME   GROUP     DRIVER                           AVAILABILITY   STATUS
hcloud-hel1             hetznercloud/hcloud-csi-driver   active         created
hcloud-nbg1             hetznercloud/hcloud-csi-driver   active         created


docker service create --name hcloud-hel1-serv   --mount type=cluster,src=hcloud-hel1,dst=/srv/www   nginx:alpine
docker service create --name hcloud-nbg1-serv   --mount type=cluster,src=hcloud-nbg1,dst=/srv/www   nginx:alpine

# hel1 serv will never work, server is in nbg1
# nbg1 is not getting created due to hel1 being stuck

# remove service hel1
docker service rm hcloud-hel1-serv

# logs still show this:
Feb 14 13:35:21 csi-dev02 dockerd[6111]: time="2023-02-14T13:35:21.732310152Z" level=info msg="error handling volume" attempt=9 error="error publishing or unpublishing to some nodes: [t7v2vpe1q0ykbjw4pmgahfnxg]" module=csi/manager node.id=t7v2vpe1q0ykbjw4pmgahfnxg volume.id=w9bokbszki1emyhxiurvyy1db
[
    {
        "ClusterVolume": {
            "ID": "w9bokbszki1emyhxiurvyy1db",
            "Version": {
                "Index": 3130
            },
            "CreatedAt": "2023-02-14T13:29:53.099835784Z",
            "UpdatedAt": "2023-02-14T13:35:21.730974867Z",
            "Spec": {
                "AccessMode": {
                    "Scope": "single",
                    "Sharing": "onewriter",
                    "MountVolume": {}
                },
                "AccessibilityRequirements": {
                    "Requisite": [
                        {
                            "Segments": {
                                "csi.hetzner.cloud/location": "hel1"
                            }
                        }
                    ]
                },
                "CapacityRange": {
                    "RequiredBytes": 11811160064,
                    "LimitBytes": 0
                },
                "Availability": "active"
            },
            "PublishStatus": [
                {
                    "NodeID": "t7v2vpe1q0ykbjw4pmgahfnxg",
                    "State": "pending-publish"
                }
            ],
            "Info": {
                "CapacityBytes": 11811160064,
                "VolumeID": "28101771",
                "AccessibleTopology": [
                    {
                        "Segments": {
                            "csi.hetzner.cloud/location": "hel1"
                        }
                    }
                ]
            }
        },
        "CreatedAt": "2023-02-14 13:29:53.099835784 +0000 UTC",
        "Driver": "hetznercloud/hcloud-csi-driver",
        "Labels": null,
        "Mountpoint": "",
        "Name": "hcloud-hel1",
        "Options": null,
        "Scope": "global"
    }
]

docker volume rm -f hcloud-hel1
# nbg service still does not schedule

# restarting docker daemon as root:
service docker restart

# scale down service and scale it up again
docker service scale hcloud-nbg1-serv=0
docker service scale hcloud-nbg1-serv=1

# remove service and add it again (doesnt work)
docker service rm hcloud-nbg1-serv
docker service create --name hcloud-nbg1-serv   --mount type=cluster,src=hcloud-nbg1,dst=/srv/www   nginx:alpine

# error log:
Feb 14 13:43:03 csi-dev02 dockerd[773]: time="2023-02-14T13:43:03Z" level=info msg="level=debug ts=2023-02-14T13:43:03.03999663Z component=grpc-server msg=\"handling request\" method=/csi.v1.Node/NodeStageVolume req=\"volume_id:\\\"28101769\\\" staging_target_path:\\\"/data/staged/tzbqvonl31mz1ulvtqbi1toq5\\\" volume_capability:<mount:<> access_mode:<mode:SINGLE_NODE_WRITER > > \"" plugin=45550d3efcd697d97d4e9e39d32b7a94d2e964cdef4ff0410433f0484c3f6350
Feb 14 13:43:03 csi-dev02 dockerd[773]: time="2023-02-14T13:43:03Z" level=info msg="level=debug ts=2023-02-14T13:43:03.040164719Z component=grpc-server msg=\"finished handling request\" method=/csi.v1.Node/NodeStageVolume err=null" plugin=45550d3efcd697d97d4e9e39d32b7a94d2e964cdef4ff0410433f0484c3f6350
Feb 14 13:43:03 csi-dev02 dockerd[773]: time="2023-02-14T13:43:03.040652761Z" level=info msg="volume staged to path /data/staged/tzbqvonl31mz1ulvtqbi1toq5" attempt=8 module=node/agent/csi volume.id=tzbqvonl31mz1ulvtqbi1toq5
Feb 14 13:43:03 csi-dev02 dockerd[773]: time="2023-02-14T13:43:03.040760708Z" level=debug msg="staging volume succeeded, attempting to publish volume" attempt=8 module=node/agent/csi volume.id=tzbqvonl31mz1ulvtqbi1toq5
Feb 14 13:43:03 csi-dev02 dockerd[773]: time="2023-02-14T13:43:03Z" level=info msg="level=debug ts=2023-02-14T13:43:03.041211408Z component=grpc-server msg=\"handling request\" method=/csi.v1.Node/NodePublishVolume req=\"volume_id:\\\"28101769\\\" staging_target_path:\\\"/data/staged/tzbqvonl31mz1ulvtqbi1toq5\\\" target_path:\\\"/data/published/tzbqvonl31mz1ulvtqbi1toq5\\\" volume_capability:<mount:<> access_mode:<mode:SINGLE_NODE_WRITER > > \"" plugin=45550d3efcd697d97d4e9e39d32b7a94d2e964cdef4ff0410433f0484c3f6350
Feb 14 13:43:03 csi-dev02 dockerd[773]: time="2023-02-14T13:43:03Z" level=info msg="level=error ts=2023-02-14T13:43:03.041281311Z component=grpc-server msg=\"handler failed\" err=\"rpc error: code = InvalidArgument desc = missing device path\"" plugin=45550d3efcd697d97d4e9e39d32b7a94d2e964cdef4ff0410433f0484c3f6350

# force remove everything that is left
docker service rm hcloud-nbg1-serv
docker volume rm -f hcloud-nbg1

# add nbg volume again (idempotency will use same volume from before)
docker volume create --driver hetznercloud/hcloud-csi-driver --required-bytes 11G --type mount --sharing onewriter --scope single hcloud-nbg1 --topology-required csi.hetzner.cloud/location=nbg1
# add service again
docker service create --name hcloud-nbg1-serv   --mount type=cluster,src=hcloud-nbg1,dst=/srv/www   nginx:alpine

# works
```

**Comment by s4ke** [comment]
@apricote could it be the case that you had a similar situation where a "bad" volume caused the proper one not being scheduled?

**Comment by apricote** [comment]
> @apricote could it be the case that you had a similar situation where a "bad" volume caused the proper one not being scheduled?

I dont think so. I only ever created one volume in the cluster in `hel1`, same as the server. My volume status is also `in use (1 node)`, but still not attached properly:

```
root@csi-test-swarm:~# docker volume ls --cluster
VOLUME NAME     GROUP     DRIVER                           AVAILABILITY   STATUS
hcloud-debug1             hetznercloud/hcloud-csi-driver   active         in use (1 node)
```

**Comment by s4ke** [comment]
Can you try restarting the docker daemon (if you still have this node running)?

**Comment by apricote** [comment]
> Can you try restarting the docker daemon (if you still have this node running)?

After restarting the volume is now marked as `published` but its not attached to the server in the API.

Also I have multiple "names" for the plugin in my config:

- Volume is managed by driver `hetznercloud/hcloud-csi-driver`
- Node has `CSIInfo` for plugins `hetznercloud/hcloud-csi-driver:latest` & `hetznercloud/hcloud-csi-driver`
- Plugin has name `hetznercloud/hcloud-csi-driver:latest`

---

I will try this whole process again tomorrow, just to make sure that I did not accidentally did something other than what I documented.

**Comment by s4ke** [comment]
Thanks! I really appreciate your help with this!

**Comment by apricote** [comment]
I tried it again today and it did not work.

I did manage to get it working by changing the driver reference to `hetznercloud/hcloud-csi-driver:latest` when creating the volume:

```diff
 docker volume create \
-  --driver hetznercloud/hcloud-csi-driver \
+  --driver hetznercloud/hcloud-csi-driver:latest \
   --required-bytes 50G \
   --type mount \
   --sharing onewriter \
   --scope single hcloud-debug1 \
   --topology-required csi.hetzner.cloud/location=hel1
```

No idea why docker starts referring to the plugin under another name than the one that I specified.

**Comment by s4ke** [comment]
Ah great. I guess this is then an issue for moby/moby.

**Review by apricote — APPROVED** [review]

**Pull request merged**